### ISSUE TITLE
Count function

### ISSUE THREAD

**Issue by thelloworldt** [litedb-org/LiteDB#1818]
Hi,

We have implemented logging using LiteDb. So I need to collect data, filter by username, action, and date, and show that data to the user. 
And next code has strange behavior
```
var count = LiteDB.Query.Or(LiteDB.Query.Contains("Action", "admin"), LiteDB.Query.Contains("UserName", "admin"));
            var btw = LiteDB.Query.Between("Stamp", new BsonValue(DateTime.Now),
                new BsonValue(DateTime.Now - TimeSpan.FromDays(30)));
            var c1=DB.Count(count);
            var c2=DB.Count(btw);
            var c3= DB.Count(LiteDB.Query.And(btw,count));
```

- `c1` works fast
- `c2` works fast
- `c3` evaluates about 10-15 mins, and locks db for that time. 

Size of DB is 13Gb with 34324 items. 
Seems like DB is doing a full scan for this request. 

Do u have any ideas why it has so much lower performance, and how to improve it?
Version is 4.1.4
Thanks for your support.

**Comment by lbnascimento** [comment]
@thelloworldt You could try running `db.Shrink()` to reduce the file size.

**Comment by thelloworldt** [comment]
> @thelloworldt You could try running `db.Shrink()` to reduce the file size.

nope, that's not a solution. Shrink removes deleted items and reduces the size. But in our case, it is a log file, so there are no removed items.

Thanks.

**Comment by lbnascimento** [comment]
@thelloworldt Do you have any performance improvement if you change the order of the operations? (`And(count,btw)` instead of `And(btw,count)`)

**Comment by sgf** [comment]
Count() will load all data into memory.

**Comment by JKamsker** [comment]
On 4.1.4, `Query.And` answers one side from an index and then checks the other side as a filter on each document. So `And(btw, count)` walks the `Stamp` index, then loads every document in that date range to evaluate the two `Contains`. At ~380 KB per document (13 GB / 34k), that's where the time goes. `c1` and `c2` on their own are answered from index keys without loading documents, which is why they were fast.

Your `Between` arguments are also reversed. It is `Between(field, start, end)` with start ≤ end. On 4.1.4, `Count(btw)` alone still works because the index is walked backwards. Inside `And`, though, the reversed range is re-checked as a filter and matches nothing, so `c3` returns 0 even when it finishes. Swapping to `And(count, btw)` returns 0 as well.

4.1.4 is no longer maintained. On v5 (current: 5.0.21), with an index on `Stamp`, the query is an index range scan plus a filter:

```csharp
col.EnsureIndex("Stamp");

var or  = Query.Or(Query.Contains("Action", "admin"), Query.Contains("UserName", "admin"));
var btw = Query.Between("Stamp", DateTime.Now.AddDays(-30), DateTime.Now);

var plan  = col.Query().Where(Query.And(btw, or)).GetPlan();
// index: INDEX RANGE SCAN(Stamp BETWEEN ...)
// filters: ($.Action LIKE "%admin%" OR $.UserName LIKE "%admin%")
var count = col.Count(Query.And(btw, or));
```

Without that index it's a full scan, and a reversed range is simply empty (count 0).

In a test with 2,000 documents of 100 KB, the corrected `c3` took ~130–260 ms on 4.1.4 and ~5–15 ms on 5.0.21. v5 still reads every document in the range, though; even `c1` alone is a full scan there. With 380 KB documents, the bigger win is keeping the large payload in a separate collection so the log documents stay small. On 4.1.4 that alone took the same count from ~260 ms to ~2 ms.

Closing this as answered — if it doesn't cover your case, comment here and we'll reopen it.
